easybuild/tools/job/slurm.py

```diff
@@ -0,0 +1,174 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+##
+"""
+Support for using Slurm as a backend for --job
+
+:author: Kenneth Hoste (Ghent University)
+"""
+import re
+from distutils.version import LooseVersion
+from vsc.utils import fancylogger
+
+from easybuild.tools.build_log import EasyBuildError, print_msg
+from easybuild.tools.config import JOB_DEPS_TYPE_ABORT_ON_ERROR, JOB_DEPS_TYPE_ALWAYS_RUN, build_option
+from easybuild.tools.job.backend import JobBackend
+from easybuild.tools.filetools import which
+from easybuild.tools.run import run_cmd
+
+
+_log = fancylogger.getLogger('slurm', fname=False)
+
+
+class Slurm(JobBackend):
+    """
+    Manage SLURM server communication and create `SlurmJob` objects.
+    """
+
+    REQ_VERSION = '17'
+
+    def __init__(self, *args, **kwargs):
+        """Constructor."""
+
+        # early check for required commands
+        for cmd in ['sbatch', 'scontrol']:
+            path = which(cmd)
+            if path is None:
+                raise EasyBuildError("Required command '%s' not found", cmd)
+
+        super(Slurm, self).__init__(*args, **kwargs)
+
+        job_deps_type = build_option('job_deps_type')
+        if job_deps_type is None:
+            job_deps_type = JOB_DEPS_TYPE_ABORT_ON_ERROR
+            self.log.info("Using default job dependency type: %s", job_deps_type)
+        else:
+            self.log.info("Using specified job dependency type: %s", job_deps_type)
+
+        if job_deps_type == JOB_DEPS_TYPE_ABORT_ON_ERROR:
+            self.job_deps_type = 'afterok'
+        elif job_deps_type == JOB_DEPS_TYPE_ALWAYS_RUN:
+            self.job_deps_type = 'afterany'
+        else:
+            raise EasyBuildError("Unknown job dependency type specified: %s", job_deps_type)
+
+    def _check_version(self):
+        """Check whether version of Slurm complies with required version."""
+        (out, _) = run_cmd("sbatch --version", trace=False)
+        slurm_ver = out.strip().split(' ')[-1]
+        self.log.info("Found Slurm version %s", slurm_ver)
+
+        if LooseVersion(slurm_ver) < LooseVersion(self.REQ_VERSION):
+            raise EasyBuildError("Found Slurm version %s, but version %s or more recent is required",
+                                 slurm_ver, self.REQ_VERSION)
+
+    def init(self):
+        """
+        Initialise the PySlurm job backend.
+        """
+        self._submitted = []
+
+    def queue(self, job, dependencies=frozenset()):
+        """
+        Add a job to the queue.
+
+        :param dependencies: jobs on which this job depends.
+        """
+        if dependencies:
+            job.job_specs['dependency'] = self.job_deps_type + ':' + ':'.join(str(d.jobid) for d in dependencies)
+
```

Comment:
For dependency jobs I think one should enforce --kill-on-invalid-dep=yes
This to avoid jobs with a dependency staying in the queue with a DependencyNeverSatisfied state.
You won't get a return result from the thus canceled jobs but they do vanish from the squeue list.

Reply:
done in 0daf74b